File: SM_MentalHealthApp.Client/Services/SignalRService.cs
using Microsoft.AspNetCore.SignalR.Client;
using System.Text.Json;

namespace SM_MentalHealthApp.Client.Services
{
    public class SignalRService : ISignalRService, IAsyncDisposable
    {
        private readonly IAuthService _authService;
        private HubConnection? _connection;
        private readonly string _hubUrl;

        public HubConnection? Connection => _connection;
        public bool IsConnected => _connection?.State == HubConnectionState.Connected;

        public event Action<CallInvitation>? OnIncomingCall;
        public event Action<ChatMessage>? OnNewMessage;
        public event Action<string>? OnCallAccepted;
        public event Action<string>? OnCallRejected;
        public event Action<string>? OnCallEnded;
        public event Action<UserStatusChange>? OnUserStatusChanged;
        public event Action<bool>? OnConnectionChanged;

        public SignalRService(IAuthService authService, HttpClient httpClient)
        {
            _authService = authService;
            // Construct hub URL from HttpClient BaseAddress
            var baseUri = httpClient.BaseAddress ?? new Uri("http://localhost:5262/");
            var port = baseUri.Port != -1 ? $":{baseUri.Port}" : "";
            _hubUrl = $"{baseUri.Scheme}://{baseUri.Host}{port}/mobilehub";
        }

        public async Task StartAsync()
        {
            if (_connection != null)
            {
                await StopAsync();
            }

            var token = _authService.Token;
            if (string.IsNullOrEmpty(token))
            {
                Console.WriteLine("No auth token available for SignalR connection");
                return;
            }

            _connection = new HubConnectionBuilder()
                .WithUrl(_hubUrl, options =>
                {
                    options.AccessTokenProvider = () => Task.FromResult(token!);
                })
                .WithAutomaticReconnect()
                .Build();

            // Set up event handlers
            SetupEventHandlers();

            try
            {
                await _connection.StartAsync();
                Console.WriteLine($"SignalR connection started successfully to {_hubUrl}");
                OnConnectionChanged?.Invoke(true);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error starting SignalR connection to {_hubUrl}: {ex.Message}");
                Console.WriteLine($"Full error: {ex}");
                OnConnectionChanged?.Invoke(false);
            }

            _connection.Reconnecting += (error) =>
            {
                Console.WriteLine($"SignalR reconnecting: {error?.Message}");
                OnConnectionChanged?.Invoke(false);
                return Task.CompletedTask;
            };

            _connection.Reconnected += (connectionId) =>
            {
                Console.WriteLine($"SignalR reconnected: {connectionId}");
                OnConnectionChanged?.Invoke(true);
                return Task.CompletedTask;
            };

            _connection.Closed += (error) =>
            {
                Console.WriteLine($"SignalR connection closed: {error?.Message}");
                OnConnectionChanged?.Invoke(false);
                return Task.CompletedTask;
            };
        }

        public async Task StopAsync()
        {
            if (_connection != null)
            {
                await _connection.StopAsync();
                await _connection.DisposeAsync();
                _connection = null;
                OnConnectionChanged?.Invoke(false);
            }
        }

        private void SetupEventHandlers()
        {
            if (_connection == null) return;

            // Incoming call
            _connection.On<object>("incoming-call", (callData) =>
            {
                try
                {
                    var json = JsonSerializer.Serialize(callData);
                    var jsonDoc = JsonDocument.Parse(json);
                    var root = jsonDoc.RootElement;

                    Console.WriteLine($"📞 SignalR: Received incoming-call event: {json}");

                    // ✅ Extract channelName if available, otherwise use callId
                    var channelName = root.TryGetProperty("channelName", out var channelNameElement)
                        ? channelNameElement.GetString()
                        : root.TryGetProperty("callId", out var callIdElement)
                            ? callIdElement.GetString()
                            : "";

                    Console.WriteLine($"📞 SignalR: Extracted channelName: {channelName}");

                    // ✅ Extract callerName directly from JSON FIRST (before deserialization)
                    string? extractedCallerName = null;
                    if (root.TryGetProperty("callerName", out var callerNameElement))
                    {
                        extractedCallerName = callerNameElement.GetString();
                        Console.WriteLine($"📞 SignalR: Extracted callerName from JSON: '{extractedCallerName}'");
                    }
                    else
                    {
                        Console.WriteLine($"⚠️ SignalR: No 'callerName' property found in JSON");
                    }

                    var call = JsonSerializer.Deserialize<CallInvitation>(json, new JsonSerializerOptions
                    {
                        PropertyNameCaseInsensitive = true
                    });

                    if (call != null)
                    {
                        // ✅ Override CallId with channel name for auto-join
                        var finalChannelName = !string.IsNullOrWhiteSpace(channelName) ? channelName : call.CallId;
                        call.CallId = finalChannelName ?? "";

                        // ✅ ALWAYS use extracted callerName if available (more reliable than deserialization)
                        if (!string.IsNullOrWhiteSpace(extractedCallerName) && extractedCallerName != "Mobile User")
                        {
                            call.CallerName = extractedCallerName;
                            Console.WriteLine($"✅ SignalR: Using extracted CallerName: '{call.CallerName}'");
                        }
                        else if (string.IsNullOrWhiteSpace(call.CallerName) || call.CallerName == "Mobile User")
                        {
                            Console.WriteLine($"⚠️ SignalR: CallerName is empty or 'Mobile User'. Deserialized value: '{call.CallerName}'");
                            Console.WriteLine($"⚠️ SignalR: Extracted value: '{extractedCallerName}'");
                            // Try to use extracted value even if it's "Mobile User" as last resort
                            if (!string.IsNullOrWhiteSpace(extractedCallerName))
                            {
                                call.CallerName = extractedCallerName;
                            }
                        }

                        // ✅ Log all properties to debug caller name issue
                        Console.WriteLine($"📞 SignalR: Final CallId (channel): {call.CallId}");
                        Console.WriteLine($"📞 SignalR: CallerId: {call.CallerId}, CallerName: '{call.CallerName}', CallType: {call.CallType}");
                        Console.WriteLine($"📞 SignalR: CallerRole: '{call.CallerRole}'");

                        if (string.IsNullOrWhiteSpace(call.CallId))
                        {
                            Console.WriteLine($"⚠️ SignalR: WARNING - Channel name is empty! This will cause issues.");
                        }

                        OnIncomingCall?.Invoke(call);
                    }
                    else
                    {
                        Console.WriteLine($"❌ SignalR: Failed to deserialize CallInvitation");
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"❌ SignalR: Error processing incoming call: {ex.Message}");
                    Console.WriteLine($"❌ SignalR: Stack trace: {ex.StackTrace}");
                }
            });

            // New message
            _connection.On<object>("new-message", (messageData) =>
            {
                try
                {
                    var json = JsonSerializer.Serialize(messageData);
                    var message = JsonSerializer.Deserialize<ChatMessage>(json, new JsonSerializerOptions
                    {
                        PropertyNameCaseInsensitive = true
                    });

                    if (message != null)
                    {
                        OnNewMessage?.Invoke(message);
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error processing new message: {ex.Message}");
                }
            });

            // Call events
            _connection.On<object>("call-accepted", (data) =>
            {
                try
                {
                    var json = JsonSerializer.Serialize(data);
                    var callData = JsonSerializer.Deserialize<Dictionary<string, object>>(json);
                    if (callData?.ContainsKey("callId") == true)
                    {
                        OnCallAccepted?.Invoke(callData["callId"].ToString() ?? "");
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error processing call accepted: {ex.Message}");
                }
            });

            _connection.On<object>("call-rejected", (data) =>
            {
                try
                {
                    var json = JsonSerializer.Serialize(data);
                    var callData = JsonSerializer.Deserialize<Dictionary<string, object>>(json);
                    if (callData?.ContainsKey("callId") == true)
                    {
                        OnCallRejected?.Invoke(callData["callId"].ToString() ?? "");
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error processing call rejected: {ex.Message}");
                }
            });

            _connection.On<object>("call-ended", (data) =>
            {
                try
                {
                    Console.WriteLine($"📞 ========== SignalRService: call-ended event received ==========");
                    var json = JsonSerializer.Serialize(data);
                    Console.WriteLine($"📞 SignalRService: Raw call-ended data: {json}");
                    var jsonDoc = JsonDocument.Parse(json);
                    var root = jsonDoc.RootElement;

                    // ✅ Prefer channelName over callId for matching
                    string? callIdentifier = null;
                    if (root.TryGetProperty("channelName", out var channelNameElement))
                    {
                        callIdentifier = channelNameElement.GetString();
                        Console.WriteLine($"📞 SignalRService: Found channelName: {callIdentifier}");
                    }
                    else if (root.TryGetProperty("callId", out var callIdElement))
                    {
                        callIdentifier = callIdElement.GetString();
                        Console.WriteLine($"📞 SignalRService: Found callId: {callIdentifier}");
                    }

                    if (!string.IsNullOrEmpty(callIdentifier))
                    {
                        Console.WriteLine($"📞 SignalRService: Invoking OnCallEnded with: {callIdentifier}");
                        OnCallEnded?.Invoke(callIdentifier);
                    }
                    else
                    {
                        Console.WriteLine($"⚠️ SignalRService: No call identifier found in call-ended message");
                    }
                    Console.WriteLine($"📞 ========== End SignalRService: call-ended event ==========");
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"❌ SignalRService: Error processing call ended: {ex.Message}");
                    Console.WriteLine($"❌ SignalRService: Stack trace: {ex.StackTrace}");
                }
            });

            // User status changes
            _connection.On<object>("user-status-changed", (statusData) =>
            {
                try
                {
                    var json = JsonSerializer.Serialize(statusData);
                    var status = JsonSerializer.Deserialize<UserStatusChange>(json, new JsonSerializerOptions
                    {
                        PropertyNameCaseInsensitive = true
                    });

                    if (status != null)
                    {
                        OnUserStatusChanged?.Invoke(status);
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error processing user status change: {ex.Message}");
                }
            });

            // Error handling
            _connection.On<string>("error", (error) =>
            {
                Console.WriteLine($"SignalR error: {error}");
            });
        }

        public async Task SendMessageAsync(int targetUserId, string message)
        {
            if (_connection?.State == HubConnectionState.Connected)
            {
                try
                {
                    await _connection.InvokeAsync("SendMessage", targetUserId, message);
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error sending message: {ex.Message}");
                }
            }
        }

        public async Task InitiateCallAsync(int targetUserId, string callType)
        {
            if (_connection?.State == HubConnectionState.Connected)
            {
                try
                {
                    await _connection.InvokeAsync("InitiateCall", targetUserId, callType);
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error initiating call: {ex.Message}");
                }
            }
        }

        public async Task AcceptCallAsync(string callId)
        {
            if (_connection?.State == HubConnectionState.Connected)
            {
                try
                {
                    await _connection.InvokeAsync("AcceptCall", callId);
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error accepting call: {ex.Message}");
                }
            }
        }

        public async Task RejectCallAsync(string callId)
        {
            if (_connection?.State == HubConnectionState.Connected)
            {
                try
                {
                    await _connection.InvokeAsync("RejectCall", callId);
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error rejecting call: {ex.Message}");
                }
            }
        }

        public async Task EndCallAsync(string callId)
        {
            if (_connection?.State == HubConnectionState.Connected)
            {
                try
                {
                    await _connection.InvokeAsync("EndCall", callId);
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error ending call: {ex.Message}");
                }
            }
        }

        public async ValueTask DisposeAsync()
        {
            await StopAsync();
        }
    }
}

File: MentalHealthMobileClean/src/services/SignalRService.js
import { HubConnectionBuilder, LogLevel } from '@microsoft/signalr';

class SignalRService {
  constructor() {
    this.connection = null;
    this.isConnected = false;
    this.listeners = {
      onMessageReceived: null,
      onIncomingCall: null,
      onCallAccepted: null,
      onCallRejected: null,
      onCallEnded: null,
      onUserStatusChanged: null,
      onConnectionStateChanged: null,
    };
  }

  async initialize(hubUrl, token) {
    try {
      // Validate hubUrl
      if (!hubUrl || typeof hubUrl !== 'string' || hubUrl.trim() === '') {
        throw new Error(`Invalid SignalR Hub URL: ${hubUrl}`);
      }
      
      const validUrl = hubUrl.trim();
      if (!validUrl.startsWith('http://') && !validUrl.startsWith('https://')) {
        throw new Error(`SignalR Hub URL must start with http:// or https://: ${validUrl}`);
      }
      
      console.log('🔗 SignalR: Initializing connection to:', validUrl);
      
      this.connection = new HubConnectionBuilder()
        .withUrl(validUrl, {
          accessTokenFactory: () => token,
        })
        .withAutomaticReconnect()
        .configureLogging(LogLevel.Information)
        .build();

      // Set up event handlers
      this.setupEventHandlers();

      // Start connection
      console.log('🔌 SignalR: Starting connection...');
      await this.connection.start();
      this.isConnected = true;
      console.log('✅ SignalR: Connected successfully!');
      console.log('✅ SignalR: Connection ID:', this.connection.connectionId);
      console.log('✅ SignalR: Connection state:', this.connection.state);
      
      if (this.listeners.onConnectionStateChanged) {
        this.listeners.onConnectionStateChanged('Connected');
      }

      return true;
    } catch (error) {
      console.error('❌ SignalR: Connection failed!');
      console.error('❌ SignalR: Error message:', error.message);
      console.error('❌ SignalR: Error stack:', error.stack);
      console.error('❌ SignalR: Full error:', JSON.stringify(error, null, 2));
      this.isConnected = false;
      
      if (this.listeners.onConnectionStateChanged) {
        this.listeners.onConnectionStateChanged('Disconnected');
      }
      
      return false;
    }
  }

  setupEventHandlers() {
    if (!this.connection) return;

    // Handle incoming messages
    this.connection.on('new-message', (message) => {
      console.log('📨 SignalR: New message received:', JSON.stringify(message, null, 2));
      console.log('📨 SignalR: Message properties:', {
        hasMessage: !!message.message,
        hasText: !!message.text,
        senderId: message.senderId,
        targetUserId: message.targetUserId,
        senderName: message.senderName
      });
      
      if (this.listeners.onMessageReceived) {
        // Handle both 'message' and 'text' property names for compatibility
        const messageText = message.message || message.text || '';
        const messageToSend = {
          ...message,
          message: messageText,
          text: messageText
        };
        console.log('📨 SignalR: Calling onMessageReceived with:', messageToSend);
        this.listeners.onMessageReceived(messageToSend);
      } else {
        console.warn('⚠️ SignalR: onMessageReceived listener is not set!');
      }
    });

    // Handle incoming calls
    this.connection.on('incoming-call', (callData) => {
      console.log('📞 SignalR: Incoming call:', callData);
      if (this.listeners.onIncomingCall) {
        this.listeners.onIncomingCall(callData);
      }
    });

    // Handle call accepted
    this.connection.on('call-accepted', (callData) => {
      console.log('✅ SignalR: Call accepted:', callData);
      if (this.listeners.onCallAccepted) {
        this.listeners.onCallAccepted(callData);
      }
    });

    // Handle call rejected
    this.connection.on('call-rejected', (callData) => {
      console.log('❌ SignalR: Call rejected:', callData);
      if (this.listeners.onCallRejected) {
        this.listeners.onCallRejected(callData);
      }
    });

    // Handle call ended
    this.connection.on('call-ended', (callData) => {
      console.log('👋 SignalR: Call ended:', callData);
      if (this.listeners.onCallEnded) {
        this.listeners.onCallEnded(callData);
      }
    });

    // Handle user status changes
    this.connection.on('user-status-changed', (statusData) => {
      console.log('👤 SignalR: User status changed:', statusData);
      if (this.listeners.onUserStatusChanged) {
        this.listeners.onUserStatusChanged(statusData);
      }
    });

    // Handle connection state changes
    this.connection.onclose((error) => {
      // Log as info, not error - connection close is normal during logout
      if (error) {
        console.log('🔌 SignalR: Connection closed (expected during logout):', error.message || error);
      } else {
        console.log('🔌 SignalR: Connection closed (expected during logout)');
      }
      this.isConnected = false;
      if (this.listeners.onConnectionStateChanged) {
        this.listeners.onConnectionStateChanged('Disconnected');
      }
    });

    this.connection.onreconnecting((error) => {
      // Automatic reconnection is expected behavior - log as info, not warning
      console.log('🔄 SignalR: Reconnecting... (this is normal if connection was temporarily lost)');
      if (error) {
        console.log('🔄 SignalR: Reconnect reason:', error.message || error);
      }
      this.isConnected = false;
      if (this.listeners.onConnectionStateChanged) {
        this.listeners.onConnectionStateChanged('Reconnecting');
      }
    });

    this.connection.onreconnected((connectionId) => {
      console.log('✅ SignalR: Reconnected successfully! (automatic reconnection working)');
      console.log('✅ SignalR: New connection ID:', connectionId);
      this.isConnected = true;
      if (this.listeners.onConnectionStateChanged) {
        this.listeners.onConnectionStateChanged('Connected');
      }
    });
  }

  // Send message
  async sendMessage(targetUserId, message) {
    if (!this.isConnected || !this.connection) {
      console.warn('⚠️ SignalR: Not connected, cannot send message');
      return false;
    }

    try {
      await this.connection.invoke('SendMessage', targetUserId, message);
      console.log('📤 SignalR: Message sent successfully');
      return true;
    } catch (error) {
      console.error('❌ SignalR: Failed to send message:', error);
      return false;
    }
  }

  // Accept call
  async acceptCall(callId) {
    if (!this.isConnected || !this.connection) {
      console.warn('⚠️ SignalR: Not connected, cannot accept call');
      return false;
    }

    try {
      await this.connection.invoke('AcceptCall', callId);
      console.log('✅ SignalR: Call accepted successfully');
      return true;
    } catch (error) {
      console.error('❌ SignalR: Failed to accept call:', error);
      return false;
    }
  }

  // Reject call
  async rejectCall(callId) {
    if (!this.isConnected || !this.connection) {
      console.warn('⚠️ SignalR: Not connected, cannot reject call');
      return false;
    }

    try {
      await this.connection.invoke('RejectCall', callId);
      console.log('❌ SignalR: Call rejected successfully');
      return true;
    } catch (error) {
      console.error('❌ SignalR: Failed to reject call:', error);
      return false;
    }
  }

  // End call
  async endCall(callId) {
    if (!this.isConnected || !this.connection) {
      console.warn('⚠️ SignalR: Not connected, cannot end call');
      return false;
    }

    try {
      await this.connection.invoke('EndCall', callId);
      console.log('👋 SignalR: Call ended successfully');
      return true;
    } catch (error) {
      console.error('❌ SignalR: Failed to end call:', error);
      return false;
    }
  }

  // Set event listeners
  setEventListener(event, callback) {
    if (this.listeners.hasOwnProperty(event)) {
      this.listeners[event] = callback;
      console.log(`✅ SignalR: Event listener set for '${event}'`);
    } else {
      console.warn(`⚠️ SignalR: Unknown event '${event}'`);
    }
  }

  // Disconnect
  async disconnect() {
    if (this.connection) {
      try {
        console.log('🔌 SignalR: Disconnecting...');
        await this.connection.stop();
        console.log('🔌 SignalR: Disconnected successfully');
      } catch (error) {
        // Log as warning, not error - disconnect errors are usually harmless
        console.warn('⚠️ SignalR: Warning during disconnect (this is usually normal):', error.message || error);
      }
      this.connection = null;
      this.isConnected = false;
    }
  }

  // Get connection state
  getConnectionState() {
    if (!this.connection) return 'Disconnected';
    return this.connection.state;
  }
}

// Export singleton instance
export default new SignalRService();
